### PR TITLE
IPv6 support

### PR DESCRIPTION
**Pull request by ranisalt**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

Rebase and some updates on #2178

Hey, it's ~~2017~~ 2022 so it's high time we implemented support for IPv6 connections. The server now listens on an IPv6 socket unless it's configured to only bind to a specific IPv4 address. The address is now stored as a varchar in the database. The internal representation of the address has been changed to the
asio-native one for simplicity.

Closes #2178

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

File: schema.sql [hunk 1]
```diff
@@ -110,7 +110,7 @@ CREATE TABLE IF NOT EXISTS `account_storage` (
 ) ENGINE=InnoDB DEFAULT CHARACTER SET=utf8;
 
 CREATE TABLE IF NOT EXISTS `ip_bans` (
-  `ip` int unsigned NOT NULL,
+	`ip` varbinary(16) NOT NULL,
```
Comment (DSpeichert) [inline schema.sql:113]: tab instead of space?
```suggestion
 `ip` varbinary(16) NOT NULL,
```

File: src/ban.cpp [hunk 2]
```diff
@@ -83,22 +84,22 @@ bool IOBan::isAccountBanned(uint32_t accountId, BanInfo& banInfo)
...
 {
-	if (clientIP == 0) {
+	if (clientIP.is_unspecified()) {
 		return false;
 	}
 
 	Database& db = Database::getInstance();
 
-	DBResult_ptr result = db.storeQuery(fmt::format("SELECT `reason`, `expires_at`, (SELECT `name` FROM `players` WHERE `id` = `banned_by`) AS `name` FROM `ip_bans` WHERE `ip` = {:d}", clientIP));
+	DBResult_ptr result = db.storeQuery(fmt::format("SELECT `reason`, `expires_at`, (SELECT `name` FROM `players` WHERE `id` = `banned_by`) AS `name` FROM `ip_bans` WHERE `ip` = INET6_ATON({:d})", db.escapeString(clientIP.to_string())));
```
Comment (DSpeichert) [inline src/ban.cpp:95]:
```suggestion
	DBResult_ptr result = db.storeQuery(fmt::format("SELECT `reason`, `expires_at`, (SELECT `name` FROM `players` WHERE `id` = `banned_by`) AS `name` FROM `ip_bans` WHERE `ip` = INET6_ATON({})", db.escapeString(clientIP.to_string())));
```
this probably isn't going to be int anymore?

File: schema.sql [hunk 3]
```diff
@@ -110,7 +110,7 @@ CREATE TABLE IF NOT EXISTS `account_storage` (
 ) ENGINE=InnoDB DEFAULT CHARACTER SET=utf8;
 
 CREATE TABLE IF NOT EXISTS `ip_bans` (
-  `ip` int unsigned NOT NULL,
+ `ip` varbinary(16) NOT NULL,
```
[the remark this answers is not in the document]
Comment (ranisalt) [reply schema.sql]: @DSpeichert :angry:

**Comment by ranisalt** [comment]
Btw let's handle https://github.com/otland/forgottenserver/pull/2178#issuecomment-281863295 separately

File: src/server.cpp [hunk 4]
```diff
@@ -158,21 +158,33 @@ void ServicePort::openAcceptor(std::weak_ptr<ServicePort> weak_service, uint16_t
...
+			address = ip::address::from_string(g_config.getString(ConfigManager::IP));
 		}
 
-		acceptor->set_option(boost::asio::ip::tcp::no_delay(true));
+		acceptor = std::make_unique<ip::tcp::acceptor>(io_service, ip::tcp::endpoint(address, serverPort));
+		if (address.is_v6()) {
+			ip::v6_only option;
+			acceptor->get_option(option);
+			if (option) {
+				option = {false};
```
Comment (DSpeichert) [inline src/server.cpp:179]: Looks like compiler doesn't like that

**Review by DSpeichert — APPROVED** [review]

**Comment by ranisalt** [comment]
I can check that it is listening on IPv6 but it's quite hard to isolate networking to force IPv6 connections. I couldn't find a way to isolate IPv6 networking and running the client and server in the same machine falls back to IPv4. My home network does not support IPv6 so running from another machine is not an option.

File: data/migrations/32.lua [hunk 5]
```diff
@@ -1,3 +1,8 @@
 function onUpdateDatabase()
-	return false
+	print("> Updating database to version 32 (IPv6 support)")
+	db.query("ALTER TABLE `players` MODIFY `lastip` varbinary(16) NOT NULL")
+	db.query("ALTER TABLE `ip_bans` MODIFY `ip` varbinary(16) NOT NULL")
+	db.query("UPDATE `players` SET `lastip` = INET6_ATON(INET_NTOA(`çastip`))")
```
Comment (EPuncker) [inline data/migrations/32.lua:5]:
```suggestion
	db.query("UPDATE `players` SET `lastip` = INET6_ATON(INET_NTOA(`lastip`))")
```

**Review by EPuncker — APPROVED** [review]
other than that minor typo, I've tested and it works just fine **backward compatibility** speaking, I don't have a way to test ipv6 😅

**Review by EPuncker — APPROVED** [review]

**Comment by ranisalt** [comment]
> What is the status of this pull request? I'm thinking in using it, and testing it because I have CGNAT ipv4 address

It _should_ be working, the conflicts are mostly formatting issues that I will fix right now

**Comment by conde2** [comment]
I placed a breakpoint in 
void Connection::accept()

But seems that Mehah OTClient is not connecting to it, but using infobyip.com/tcpportchecker.php it does calls my breakpoint.

**Review by EPuncker — APPROVED** [review]

File: data/migrations/34.lua [hunk 6]
```diff
@@ -1,3 +1,8 @@
 function onUpdateDatabase()
-	return false
```
Comment (marmichalski) [inline data/migrations/34.lua:2]: Missing version bump in schema 🙈

**Pull request merged**